### PR TITLE
Cookie

### PR DESCRIPTION
**Pull request by 00salmon**
## Description of the change

> Description here

## Type of change
- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation or Development tools (readme, specs, tests, code formatting)

## Links

 - Jira issue number: DEV-2777
 - Process.st launch checklist: https://app.process.st/checklists/Safari-Cookie-Patch-lSs1NdHohAuN1T-AxBhK6g/

### Development

- [ ] [Prettier](https://www.npmjs.com/package/prettier) was run
- [ ] The behaviour changes in the pull request are covered by specs
- [ ] All tests related to the changed code pass in development

### Paperwork

- [ ] This pull request has a descriptive title and information useful to a reviewer
- [ ] This pull request has a Jira number
- [ ] This pull request has a Process.st launch checklist
- [ ] "Ready for review" label attached to the PR and reviewers pinged on Slack

### Code review 

- [ ] Changes have been reviewed by at least one other engineer
- [ ] Security impacts of this change have been considered

### CODE REVIEW

File: src/types.ts [hunk 1]
```diff
@@ -18,6 +18,7 @@ export interface WidgetConfig {
   user: User; // The user details
   widgetType?: WidgetType; // The content of the widget.
   engagementMedium?: EngagementMedium; // How to display the widget.
+  cookie?: string;
```
Comment (loganvolkers) [inline src/types.ts:21]: This is incorrect, it should be `user.cookie`
```suggestion
```

File: src/widgets/Widget.ts [hunk 2]
```diff
@@ -230,12 +231,14 @@ export default abstract class Widget {
...
         id: this.context.user.id,
         accountId: this.context.user.accountId,
+        cookie: Cookies.get("_saasquatch")
       };
 
       response = this.widgetApi.upsertUser({
         user: userObj,
         engagementMedium,
         widgetType: this.type,
+        cookie: Cookies.get("_saasquatch"),
```
Comment (loganvolkers) [inline src/widgets/Widget.ts:241]: Only needed inside the user object, not here

```suggestion
```

File: src/widgets/Widgets.ts [hunk 3]
```diff
@@ -114,6 +115,10 @@ export default class Widgets {
   async upsertUser(config: WidgetConfig) {
     const raw = config as unknown;
     const clean = validateWidgetConfig(raw);
+    // clean.user.cookie = readCookie("saasquatchCookie");
+
+    console.log("about to pass cookie to user upsert", Cookies.get("_saasquatch"))
+
```
Comment (loganvolkers) [inline src/widgets/Widgets.ts:121]: Intentional commenting and logging?

File: package.json [hunk 4]
```diff
@@ -35,26 +35,39 @@
...
+    "@babel/register": "^7.10.1",
     "@babel/runtime-corejs2": "^7.9.6",
+    "@types/chai": "^4.2.11",
+    "@types/cucumber": "^6.0.1",
     "@types/debug": "4.1.5",
+    "@types/js-cookie": "^2.2.6",
+    "@types/jsdom-global": "^3.0.2",
     "@types/react": "^16.9.35",
     "@types/superagent": "^4.1.7",
+    "@ungap/url-search-params": "^0.2.0",
```
Comment (loganvolkers) [inline package.json:51]: This is not actually used, don't need it here.

File: src/index.ts [hunk 5]
```diff
@@ -143,3 +147,53 @@ export function submitEmail(fn: (target, widget, email) => any): void {
...
+        _log("cookie to store:", newCookie);
+      } else {
+        reEncodedCookie = b64encode(JSON.stringify(paramsJSON));
+        _log("cookie to store:", paramsJSON);
+      }
+      Cookies.set("_saasquatch", reEncodedCookie, {
+        expires: 365,
+        secure: true,
+        sameSite: "lax",
+        domain,
```
Comment (loganvolkers) [inline src/index.ts:192]: We should extract this as a function such as `dropCookie`, and expose it so you can call `squatch.dropCookie()` for example when `SaaSquatchDoNotAutoDrop` is true.

File: src/index.ts [hunk 6]
```diff
@@ -143,3 +147,53 @@ export function submitEmail(fn: (target, widget, email) => any): void {
 }
 
 if (window) asyncLoad();
```
Comment (loganvolkers) [inline src/index.ts:149]: More isometric way of checking this.

```suggestion
if (typeof document !== "undefined") asyncLoad();
```

File: src/index.ts [hunk 7]
```diff
@@ -143,3 +147,53 @@ export function submitEmail(fn: (target, widget, email) => any): void {
...
+
+if (window && !window.SaaSquatchDoNotAutoDrop) {
```
Comment (loganvolkers) [inline src/index.ts:151]: More isometric was of checking this

```suggestion
if (typeof document !== "undefined" && !window.SaaSquatchDoNotAutoDrop) {
```

File: src/index.ts [hunk 8]
```diff
@@ -142,4 +143,12 @@ export function submitEmail(fn: (target, widget, email) => any): void {
   widgets().submitEmail(fn);
 }
 
-if (window) asyncLoad();
+export function dropCookie():void {
```
Comment (loganvolkers) [inline src/index.ts:146]: In retrospect "drop" sounds too much like "delete"
```suggestion
export function pushCookie():void {
```

File: src/widgets/Widget.ts [hunk 9]
```diff
@@ -230,6 +231,7 @@ export default abstract class Widget {
         lastName: lastName || null,
         id: this.context.user.id,
         accountId: this.context.user.accountId,
+        cookie: Cookies.get("_saasquatch")
```
Comment (johanventer) [inline src/widgets/Widget.ts:234]:
```suggestion
        cookies: Cookies.get("_saasquatch")
```

File: src/widgets/Widgets.ts [hunk 10]
```diff
@@ -112,7 +113,14 @@ export default class Widgets {
    * @return {Promise<WidgetResult>} json object if true, with a Widget and user details.
    */
   async upsertUser(config: WidgetConfig) {
-    const raw = config as unknown;
+    const cookieConfig = {
+      ...config,
+      user: {
+        ...config.user,
+        cookie: Cookies.get("_saasquatch")
```
Comment (johanventer) [inline src/widgets/Widgets.ts:120]:
```suggestion
        cookies: Cookies.get("_saasquatch")
```

File: src/utils/cookieUtils.ts [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+        const newCookie = deepMerge(existingCookie, paramsJSON);
+        reEncodedCookie = b64encode(JSON.stringify(newCookie));
+        _log("cookie to store:", newCookie);
+      } else {
+        reEncodedCookie = b64encode(JSON.stringify(paramsJSON));
+        _log("cookie to store:", paramsJSON);
+      }
+      Cookies.set("_saasquatch", reEncodedCookie, {
+        expires: 365,
+        secure: document.location.protocol === "https:" ? true : false,
```
Comment (johanventer) [inline src/utils/cookieUtils.ts:88]: I think we decided to always set this to false.
```suggestion
        secure: false,
```

File: src/utils/io.ts [hunk 12]
```diff
@@ -10,7 +11,7 @@ export function doGet(url, jwt = "") {
   if (jwt) headers["X-SaaSquatch-User-Token"] = jwt;
 
   const request = superagent.get(url).withCredentials().set(headers);
-
+  request.set('cookie', '_saasquatch=' + Cookies.get('_saasquatch'));
```
Comment (johanventer) [inline src/utils/io.ts:14]: Are we meant to be setting this as a header as well??

Comment (00salmon) [reply src/utils/io.ts:14]: forgot to remove this, good call

**Comment by loganvolkers** [comment]
@jorgecgll Don't forget you'll need a new pull request for the actual 2.2.0 release

**Pull request merged**